File: shadowdroid-plugin/src/main/java/io/github/andriyo/shadowdroid/studio/BridgeProtocol.java
package io.github.andriyo.shadowdroid.studio;

import com.sun.net.httpserver.HttpExchange;

import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

final class BridgeProtocol {
    static final int DEFAULT_DEBUGGER_TIMEOUT_MS = 2_500;

    private BridgeProtocol() {
    }

    static Response ok(Object... fields) {
        return new Response(HttpURLConnection.HTTP_OK, obj(fields));
    }

    static Response bad(String message) {
        return new Response(HttpURLConnection.HTTP_BAD_REQUEST, obj("ok", false, "error", message));
    }

    static void send(HttpExchange exchange, int status, String body) {
        try {
            byte[] bytes = body.getBytes(StandardCharsets.UTF_8);
            exchange.getResponseHeaders().set("content-type", "application/json; charset=utf-8");
            exchange.sendResponseHeaders(status, bytes.length);
            try (var out = exchange.getResponseBody()) {
                out.write(bytes);
            }
        } catch (IOException ignored) {
        }
    }

    static Map<String, String> parseQuery(String raw) {
        if (raw == null || raw.isBlank()) return Collections.emptyMap();
        Map<String, String> params = new LinkedHashMap<>();
        for (String part : raw.split("&")) {
            int index = part.indexOf('=');
            if (index < 0) continue;
            params.put(decode(part.substring(0, index)), decode(part.substring(index + 1)));
        }
        return params;
    }

    static int intParam(Map<String, String> query, String key, int defaultValue, int min, int max) {
        String value = query.get(key);
        if (value == null) return defaultValue;
        try {
            int parsed = Integer.parseInt(value);
            return Math.max(min, Math.min(max, parsed));
        } catch (NumberFormatException ignored) {
            return defaultValue;
        }
    }

    static int debuggerTimeoutMs(Map<String, String> query) {
        return intParam(query, "timeout_ms", DEFAULT_DEBUGGER_TIMEOUT_MS, 50, 30_000);
    }

    static boolean booleanParam(Map<String, String> query, String key, boolean defaultValue) {
        String value = query.get(key);
        return value == null ? defaultValue : Boolean.parseBoolean(value);
    }

    static long nowMs() {
        return System.currentTimeMillis();
    }

    static Map<String, Object> map(Object... fields) {
        Map<String, Object> map = new LinkedHashMap<>();
        for (int i = 0; i + 1 < fields.length; i += 2) {
            map.put(fields[i].toString(), fields[i + 1]);
        }
        return map;
    }

    static String obj(Object... fields) {
        return json(map(fields));
    }

    @SuppressWarnings("unchecked")
    static String json(Object value) {
        if (value == null) return "null";
        if (value instanceof String string) return "\"" + escape(string) + "\"";
        if (value instanceof Number || value instanceof Boolean) return value.toString();
        if (value instanceof Map<?, ?> map) {
            StringBuilder builder = new StringBuilder("{");
            boolean first = true;
            for (Map.Entry<?, ?> entry : map.entrySet()) {
                if (!first) builder.append(',');
                first = false;
                builder.append(json(String.valueOf(entry.getKey()))).append(':').append(json(entry.getValue()));
            }
            return builder.append('}').toString();
        }
        if (value instanceof Iterable<?> iterable) {
            StringBuilder builder = new StringBuilder("[");
            boolean first = true;
            for (Object item : iterable) {
                if (!first) builder.append(',');
                first = false;
                builder.append(json(item));
            }
            return builder.append(']').toString();
        }
        if (value.getClass().isArray()) {
            List<Object> list = new ArrayList<>();
            Object[] array = (Object[]) value;
            Collections.addAll(list, array);
            return json(list);
        }
        return json(value.toString());
    }

    private static String decode(String value) {
        return URLDecoder.decode(value, StandardCharsets.UTF_8);
    }

    private static String escape(String value) {
        StringBuilder builder = new StringBuilder(value.length() + 16);
        for (int i = 0; i < value.length(); i++) {
            char c = value.charAt(i);
            switch (c) {
                case '\\' -> builder.append("\\\\");
                case '"' -> builder.append("\\\"");
                case '\n' -> builder.append("\\n");
                case '\r' -> builder.append("\\r");
                case '\t' -> builder.append("\\t");
                default -> {
                    if (c < 0x20) builder.append(String.format("\\u%04x", (int) c));
                    else builder.append(c);
                }
            }
        }
        return builder.toString();
    }
}

File: shadowdroid-plugin/src/main/kotlin/io/github/andriyo/shadowdroid/studio/BridgeProtocol.kt
package io.github.andriyo.shadowdroid.studio

import com.sun.net.httpserver.HttpExchange
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URLDecoder
import java.nio.charset.StandardCharsets
import kotlin.math.max
import kotlin.math.min

internal object BridgeProtocol {
    const val DEFAULT_DEBUGGER_TIMEOUT_MS: Int = 2_500

    @JvmStatic
    fun ok(vararg fields: Any?): Response = Response(HttpURLConnection.HTTP_OK, obj(*fields))

    @JvmStatic
    fun bad(message: String?): Response =
        Response(HttpURLConnection.HTTP_BAD_REQUEST, obj("ok", false, "error", message))

    @JvmStatic
    fun send(exchange: HttpExchange, status: Int, body: String) {
        try {
            val bytes = body.toByteArray(StandardCharsets.UTF_8)
            exchange.responseHeaders["content-type"] = "application/json; charset=utf-8"
            exchange.sendResponseHeaders(status, bytes.size.toLong())
            exchange.responseBody.use { it.write(bytes) }
        } catch (_: IOException) {
        }
    }

    @JvmStatic
    fun parseQuery(raw: String?): Map<String, String> {
        if (raw.isNullOrBlank()) return emptyMap()
        val params = linkedMapOf<String, String>()
        for (part in raw.split("&")) {
            val index = part.indexOf('=')
            if (index < 0) continue
            params[decode(part.substring(0, index))] = decode(part.substring(index + 1))
        }
        return params
    }

    @JvmStatic
    fun intParam(query: Map<String, String>, key: String, defaultValue: Int, min: Int, max: Int): Int {
        val value = query[key] ?: return defaultValue
        return value.toIntOrNull()?.let { parsed -> max(min, min(max, parsed)) } ?: defaultValue
    }

    @JvmStatic
    fun debuggerTimeoutMs(query: Map<String, String>): Int =
        intParam(query, BridgeQuery.TIMEOUT_MS, DEFAULT_DEBUGGER_TIMEOUT_MS, 50, 30_000)

    @JvmStatic
    fun booleanParam(query: Map<String, String>, key: String, defaultValue: Boolean): Boolean =
        query[key]?.toBoolean() ?: defaultValue

    @JvmStatic
    fun nowMs(): Long = System.currentTimeMillis()

    @JvmStatic
    fun map(vararg fields: Any?): MutableMap<String, Any?> {
        val payload = linkedMapOf<String, Any?>()
        var index = 0
        while (index + 1 < fields.size) {
            payload[fields[index].toString()] = fields[index + 1]
            index += 2
        }
        return payload
    }

    @JvmStatic
    fun obj(vararg fields: Any?): String = json(map(*fields))

    @JvmStatic
    fun json(value: Any?): String = when (value) {
        null -> "null"
        is String -> "\"${escape(value)}\""
        is Number, is Boolean -> value.toString()
        is Map<*, *> -> value.entries.joinToString(prefix = "{", postfix = "}") { (key, item) ->
            "${json(key.toString())}:${json(item)}"
        }
        is Iterable<*> -> value.joinToString(prefix = "[", postfix = "]") { json(it) }
        is Array<*> -> value.joinToString(prefix = "[", postfix = "]") { json(it) }
        else -> json(value.toString())
    }

    private fun decode(value: String): String = URLDecoder.decode(value, StandardCharsets.UTF_8)

    private fun escape(value: String): String = buildString(value.length + 16) {
        for (char in value) {
            when (char) {
                '\\' -> append("\\\\")
                '"' -> append("\\\"")
                '\n' -> append("\\n")
                '\r' -> append("\\r")
                '\t' -> append("\\t")
                else -> {
                    if (char < ' ') append("\\u%04x".format(char.code)) else append(char)
                }
            }
        }
    }
}
